funsor/optimizer.py

```diff
@@ -0,0 +1,224 @@
+from __future__ import absolute_import, division, print_function
+
+import collections
+from six.moves import reduce
+
+from opt_einsum.paths import greedy
+
+from funsor.domains import find_domain
+from funsor.interpreter import interpretation, reinterpret
+from funsor.ops import ASSOCIATIVE_OPS, DISTRIBUTIVE_OPS
+from funsor.registry import KeyedRegistry
+from funsor.terms import eager, reflect, Binary, Funsor, Reduce
+
+
+class Finitary(Funsor):
+    """
+    Lazy finitary operation. Used internally in the optimizer.
+    Finitary(op, operands) == six.moves.reduce(op, operands)
+    """
+    def __init__(self, op, operands):
+        assert callable(op)
+        assert isinstance(operands, tuple)
+        assert all(isinstance(operand, Funsor) for operand in operands)
+        inputs = collections.OrderedDict()
+        for operand in operands:
+            inputs.update(operand.inputs)
+
+        output = reduce(lambda lhs, rhs: find_domain(op, lhs, rhs),
+                        [operand.output for operand in reversed(operands)])
+
+        super(Finitary, self).__init__(inputs, output)
+        self.op = op
+        self.operands = operands
+
+    def __repr__(self):
+        return 'Finitary({}, {})'.format(self.op.__name__, self.operands)
+
+    def eager_subs(self, subs):
+        if not any(k in self.inputs for k, v in subs):
+            return self
+        operands = tuple(operand.eager_subs(subs) for operand in self.operands)
+        return Finitary(self.op, operands)
+
+
+@eager.register(Finitary, object, tuple)
+def eager_finitary(op, operands):
+    return reduce(lambda lhs, rhs: Binary(op, lhs, rhs), operands)
+
+
+def associate(cls, *args):
+    result = _associate(cls, *args)
+    if result is None:
+        result = reflect(cls, *args)
+    return result
+
+
+_associate = KeyedRegistry(default=lambda *args: None)
+associate.register = _associate.register
+
+
+@associate.register(Binary, object, Funsor, Funsor)
+def binary_to_finitary(op, lhs, rhs):
+    """convert Binary to Finitary"""
+    return Finitary(op, (lhs, rhs))
+
+
+@associate.register(Finitary, object, tuple)
+def associate_finitary(op, operands):
+    # Finitary(Finitary) -> Finitary
+    new_operands = []
+    for term in operands:
+        if isinstance(term, Finitary) and term.op is op and op in ASSOCIATIVE_OPS:
+            new_operands.extend(term.operands)
+        else:
+            new_operands.append(term)
+
+    with interpretation(reflect):
```

Comment:
So far I've been able to avoid changing interpretation inside constructors by using logic like
```py
def associate_finitary(op, operands):
    if any(isinstance(x, Finitary) for x in operands):
        # Eagerly fuse Finitary terms.
        new_operands = []
        ...
        return Finitary(op, new_operands)

    return None  # defer to default implementation
```
This seems a little safer, but I don't feel strongly either way.

I'll merge now, but feel free to send a quick follow-up PR if you prefer the other style.